photon-api/src/main/scala/com/linkedin/photon/ml/data/LocalDataSet.scala

```diff
@@ -212,6 +239,104 @@ object LocalDataSet {
 
+  /**
+    *  Compute Ratio Confidence Interval lower bounds.
+    *
+    *  @param randomLabelAndFeatures An array of (label, feature) tuples
+    *  @param quartile The quartile score of standard normal distribution
+    *  @param globalFeatureShardStats The global population feature statistics to help compute ratio
+    */
+
+  protected[ml] def computeRatioCILowerBound(
+    randomLabelAndFeatures: Array[(Double, Vector[Double])],
+    quartile: Double,
+    globalFeatureShardStats: BasicStatisticalSummary): Map[Int, Double] = {
+
+    val dummyBoundForNonBinaryOrIntercept = 2.0
+    val randomFeatureFirstOrderSums = mutable.Map[Int, Double]()
+    val lowerBounds = mutable.Map[Int, Double]()
+    var randomEffectNumSamples = 0
+    var globalPopulationNumSamples = globalFeatureShardStats.count
+    val globalPopulationFirstOrderMeans = globalFeatureShardStats.mean.toArray
+    val globallFeatureNonZero = globalFeatureShardStats.numNonzeros.toArray
+    val globalMean = globalFeatureShardStats.mean.toArray
+    randomLabelAndFeatures.foreach { case (label, features) =>
+      randomEffectNumSamples += 1
+     features.activeIterator.foreach { case (key, value) =>
+       randomFeatureFirstOrderSums.update(key, randomFeatureFirstOrderSums.getOrElse(key, 0.0) + value)}
+    }
+
+    val m: Int = randomEffectNumSamples
+    val n: Long = globalPopulationNumSamples
+    val lastColumn = randomFeatureFirstOrderSums.keySet.max
+
+    randomFeatureFirstOrderSums.keySet.foreach { key =>
+      // Do computation on only binary and non-intercept features
+      if ((globallFeatureNonZero(key) * 1.0) / globalPopulationNumSamples == globalMean(key) && key != lastColumn) {
+        var x: Double = randomFeatureFirstOrderSums.get(key).get
+        var y: Double = globalPopulationFirstOrderMeans(key) * globalPopulationNumSamples
+        var lowerBound = None: Option[Double]
+        var upperBound = None: Option[Double]
+        var py = None: Option[Double]
+        // deal with extreme cases of x and y
+        (x, y) match {
+          case d if (d._1 == 0 && d._2 == 0) => {
+            lowerBound = Some(0.0)
+            upperBound = Some(Double.PositiveInfinity)
+          }
+          case d if (d._1 == 0 && d._2 != 0) => {
+            lowerBound = Some(0.0)
+            x = 0.5
+          }
+          case d if (d._1 != 0 && d._2 == 0) => {
+            y = 0.5
+            py = Some(y / n)
+            upperBound = Some(Double.PositiveInfinity)
+          }
+          case d if (d._1 == m && d._2 == n) => {
+            x = m - 0.5
+            y = n - 0.5
+            py = Some(y / n)
+          }
+          case _ => None
+        }
+        // we have the mean already from the statistics, save computation here
+        py match {
+          case Some(py) => None
+          case _ => py = Some(globalPopulationFirstOrderMeans(key))
+        }
+        val t = ( x / m ) / py.get
+        val variance = 1.0 / x - 1.0 / m + 1.0 / y - 1.0 / n
```

Comment:
We should extract this out to a function that returns (mean, variance). There's too much going on here in-line.